hugr-core/src/envelope.rs

```diff
@@ -319,14 +319,77 @@ fn encode_model<'h>(
 }
 
 #[cfg(test)]
-mod tests {
+pub(crate) mod test {
     use super::*;
     use cool_asserts::assert_matches;
     use rstest::rstest;
+    use std::borrow::Cow;
     use std::io::BufReader;
 
     use crate::builder::test::{multi_module_package, simple_package};
     use crate::extension::PRELUDE_REGISTRY;
+    use crate::hugr::test::check_hugr_equality;
+    use crate::std_extensions::STD_REG;
+    use crate::HugrView;
+
+    /// Returns an `ExtensionRegistry` with the extensions from both
+    /// sets. Avoids cloning if the first one already contains all
+    /// extensions from the second one.
+    fn join_extensions<'a>(
+        extensions: &'a ExtensionRegistry,
+        other: &ExtensionRegistry,
+    ) -> Cow<'a, ExtensionRegistry> {
+        if other.iter().all(|e| extensions.contains(e.name())) {
+            Cow::Borrowed(extensions)
+        } else {
+            let mut extensions = extensions.clone();
+            extensions.extend(other);
+            Cow::Owned(extensions)
+        }
+    }
+
+    /// Serialize and deserialize a HUGR into a binary envelope, and check that the
+    /// result is the same as the original.
+    ///
+    /// We do not compare the before and after `Hugr`s for
+    /// equality directly, because impls of `CustomConst` are not required to implement
+    /// equality checking.
+    ///
+    /// Returns the deserialized HUGR.
+    #[expect(unused)]
+    pub(crate) fn check_hugr_bin_roundtrip(hugr: &Hugr) -> Hugr {
+        let mut buffer = Vec::new();
+        hugr.store(&mut buffer, EnvelopeConfig::binary()).unwrap();
+
+        let extensions = join_extensions(&STD_REG, hugr.extensions());
+
+        let reader = BufReader::new(buffer.as_slice());
+        let extracted = Hugr::load(reader, Some(&extensions)).unwrap();
+
+        check_hugr_equality(&extracted, hugr);
+        extracted
+    }
+
+    /// Serialize and deserialize a HUGR into a text envelope, and check that the
+    /// result is the same as the original.
+    ///
+    /// We do not compare the before and after `Hugr`s for
+    /// equality directly, because impls of `CustomConst` are not required to implement
+    /// equality checking.
+    ///
+    /// Returns the deserialized HUGR.
+    pub(crate) fn check_hugr_text_roundtrip(hugr: &Hugr) -> Hugr {
```

Comment:
nit: both of these could call a shared implementation that takes an envelope config as function argument